src/circuit_writer/mod.rs

```diff
@@ -62,17 +61,12 @@ impl<B: Backend> CircuitWriter<B> {
         self.typed.expr_type(expr)
     }
 
-    // TODO: can we get rid of this?
-    pub fn node_type(&self, node_id: usize) -> Option<&TyKind> {
-        self.typed.node_type(node_id)
-    }
-
     pub fn struct_info(&self, qualified: &FullyQualified) -> Option<&StructInfo> {
         self.typed.struct_info(qualified)
     }
 
-    pub fn fn_info(&self, qualified: &FullyQualified) -> Option<&FnInfo<B>> {
-        self.typed.fn_info(qualified)
+    pub fn fn_info(&self, expr: &Expr) -> Option<&FnInfo<B>> {
```

Comment:
It is intended to align with the `CircuitWriter::expr_type`

Alternatively, we could store the instantiated function ast via its function signature, in which the arguments are monomorphized and in concrete type.

Performance-wise, the alternative approach would be better as it can reuse a function instantiation if the arguments are the same.